### PR TITLE
Fix infant anatomical ingression

### LINKED ISSUE #124: —infant results in pipeline flagging missing adult T1w QSIPrep outputs
## Summary
I am trying to run qsirecon with the infant pipeline, but nipype.workflow flags the adult MNI anatomical outputs as missing.

I suspect it’s related to https://github.com/PennLINC/qsirecon/issues/44?

## Additional details

- QSIRecon version: 0.23.3.dev0+g7b29424.d20241016
- Docker version:
- Singularity version: apptainer version 1.3.2-1.el7

### What were you trying to do?
Run qsirecon with one of the built-in reconstruction workflows.

### What did you expect to happen?
QSIrecon to run, since https://github.com/PennLINC/qsirecon/blob/7b29424d55f3b13bb12865e9a51184f5f0358873/qsirecon/workflows/base.py#L249
suggests that there is a switch built in for —infant with the MNI Infant template.

### What actually happened?
```
241017-15:24:19,696 nipype.workflow INFO:
	 found ['/scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/dwi/sub-16001_acq-b700_dir-PA_space-T1w_desc-preproc_dwi.nii.gz'] in /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24
241017-15:24:19,698 nipype.workflow INFO:
	 CHECKING /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_desc-brain_mask.nii.gz: True
241017-15:24:19,698 nipype.workflow INFO:
	 CHECKING /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_desc-preproc_T1w.nii.gz: True
241017-15:24:19,698 nipype.workflow INFO:
	 Found usable QSIPrep-preprocessed T1w image and mask.
241017-15:24:19,698 nipype.workflow INFO:
	 CHECKING /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_from-T1w_to-MNI152NLin2009cAsym_mode-image_xfm.h5: False
241017-15:24:19,698 nipype.workflow INFO:
	 CHECKING /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_from-T1w_to-MNI152NLin2009cAsym_mode-image_xfm.h5: False
241017-15:24:19,698 nipype.workflow INFO:
	 CHECKING /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_from-MNI152NLin2009cAsym_to-T1w_mode-image_xfm.h5: False
241017-15:24:19,698 nipype.workflow INFO:
	 CHECKING /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_from-MNI152NLin2009cAsym_to-T1w_mode-image_xfm.h5: False
241017-15:24:19,698 nipype.workflow INFO:
	 Missing T1w QSIPrep outputs: /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_from-T1w_to-MNI152NLin2009cAsym_mode-image_xfm.h5 /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24/sub-16001/anat/sub-16001_from-MNI152NLin2009cAsym_to-T1w_mode-image_xfm.h5
Process Process-2:
Traceback (most recent call last):
  File "/opt/conda/envs/qsiprep/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/opt/conda/envs/qsiprep/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon/cli/workflow.py", line 139, in build_workflow
    retval["workflow"] = init_qsirecon_wf()
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon/workflows/base.py", line 43, in init_qsirecon_wf
    single_subject_wf = init_single_subject_recon_wf(subject_id=subject_id)
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon/workflows/base.py", line 130, in init_single_subject_recon_wf
    anat_ingress_node, available_anatomical_data = init_highres_recon_anatomical_wf(
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon/workflows/recon/anatomical.py", line 87, in init_highres_recon_anatomical_wf
    raise Exception("Cannot compute to-template")
Exception: Cannot compute to-template
```

## Reproducing the bug
```
export APPTAINERENV_FS_LICENSE=$GROUP_HOME/freesurferlicense.txt
singularity run --containall --writable-tmpfs --cleanenv \
/home/groups/jyeatman/software/singularity_images/qsirecon-0.23.2.sif \
/scratch/groups/jyeatman/howard/DBPFullwT1derivOct24 \
/scratch/groups/jyeatman/howard/DBPFullwT1derivOct24-recon/ participant --participant-label sub-$SLURM_ARRAY_TASK_ID \
-w /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24-work \
--infant \
--recon-spec mrtrix_singleshell_ss3t_noACT \
--fs-license-file $GROUP_HOME/freesurferlicense.txt \
--nthreads 10 \
--omp-nthreads 8 \
--output-resolution 2 \
-v -v
```

### PR DESCRIPTION
**Pull request by tsalo**
Closes #124.

## Changes proposed in this pull request

- Grab transforms to/from MNIInfant instead of MNI152NLin2009cAsym if infant mode is enabled.

### CODE REVIEW

File: qsirecon/workflows/recon/anatomical.py [hunk 1]
```diff
@@ -51,10 +51,17 @@
...
-QSIRECON_NORMALIZED_ANAT_REQUIREMENTS = [
-    "sub-{subject_id}/anat/sub-{subject_id}_from-T1w_to-MNI152NLin2009cAsym_mode-image_xfm.h5",
-    "sub-{subject_id}/anat/sub-{subject_id}_from-MNI152NLin2009cAsym_to-T1w_mode-image_xfm.h5",
-]
+if config.workflow.infant:
+    # XXX: Not robust to different cohorts...
+    QSIRECON_NORMALIZED_ANAT_REQUIREMENTS = [
+        "sub-{subject_id}/anat/sub-{subject_id}_from-T1w_to-MNIInfant_mode-image_xfm.h5",
+        "sub-{subject_id}/anat/sub-{subject_id}_from-MNIInfant_to-T1w_mode-image_xfm.h5",
+    ]
```
[the remark this answers is not in the document]
Comment (tsalo) [reply qsirecon/workflows/recon/anatomical.py:59]: Also probably important to keep an eye toward session-wise processing.

[the remark this answers is not in the document]
Comment (chiuhoward) [reply qsirecon/workflows/recon/anatomical.py:59]:
```
sub-{subject_id}/anat/sub-{subject_id}_from-MNIInfant_to-T1w_mode-image_xfm.h5
sub-{subject_id}/anat/sub-{subject_id}_from-T1w_to-MNIInfant_mode-image_xfm.h5
```

if it helps, @tsalo these are the filenames for my qsiprep h5 output files (single session data)

[the remark this answers is not in the document]
Comment (tsalo) [reply qsirecon/workflows/recon/anatomical.py:59]: Thanks @chiuhoward. I do worry that that naming convention will change once we address https://github.com/PennLINC/qsiprep/issues/826. 

I also took a look at some nibabies derivatives and the transforms have the following names:

```
sub-<label>_ses-<label>_run-<index>_from-MNIInfant+1_to-T1w_mode-image_xfm.h5
sub-<label>_ses-<label>_run-<index>_from-T1w_to-MNIInfant+1_mode-image_xfm.h5
```

[the remark this answers is not in the document]
Comment (tsalo) [reply qsirecon/workflows/recon/anatomical.py:59]: Still, we can merge this fix and just open an issue about handling different cohorts in the future.

**Comment by tsalo** [comment]
@chiuhoward would you be able to check out my branch and test it on your infant data? Otherwise it'll be a couple of days before I can hunt down some infant data and actually run tests.

**Comment by chiuhoward** [comment]
@tsalo sure let me try and get back to you - i usually run it from a docker build but will test it in a conda environment

**Comment by tsalo** [comment]
I didn't change any dependencies so you should be able to just mount your cloned copy of the repo into the Docker image.

E.g.,

```
docker run --rm -ti \
    -v /path/to/qsirecon:/usr/local/miniconda/lib/python3.10/site-packages/qsirecon \
    pennlinc/qsirecon:unstable \
    ...
```

**Comment by tsalo** [comment]
@chiuhoward I don't know if you've tried running your test yet, but the mounting path was wrong.

It should be:

```
docker run --rm -ti \
    -v /path/to/qsirecon:/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon \
    pennlinc/qsirecon:unstable \
    ...
```

**Comment by chiuhoward** [comment]
@tsalo I’ll try that today thank you!

**Comment by chiuhoward** [comment]
@tsalo I tried running the patch with two subjects and hit another error with `atlas_config.json`. Will be happy to email you the crash logs if they’ll be helpful!

```
 Saving crash info to /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24-recon/sub-16004/log/20241025-150355_fed16825-5a08-4cee-84fb-cb68921d3e53/crash-20241025-150430-chiuh-get_atlases-1b41c934-2609-4559-b89a-c5100efcec5b.txt
Traceback (most recent call last):
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/nipype/pipeline/plugins/multiproc.py", line 344, in _send_procs_to_workers
    self.procs[jobid].run(updatehash=updatehash)
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node get_atlases.

Traceback:
	Traceback (most recent call last):
	  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/nipype/interfaces/base/core.py", line 397, in run
	    runtime = self._run_interface(runtime)
	  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon/interfaces/utils.py", line 48, in _run_interface
	    atlas_configs = get_atlases(atlas_names)
	  File "/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon/utils/atlases.py", line 45, in get_atlases
	    with open(os.path.join(atlas_dir, "atlas_config.json")) as f:
	FileNotFoundError: [Errno 2] No such file or directory: '/atlas/qsirecon_atlases/atlas_config.json'
```

I was running this after cloning the fix-infant branch:

```
singularity run --containall --writable-tmpfs --cleanenv \
--bind /home/groups/jyeatman/software/howard/qsirecon/qsirecon/qsirecon:/opt/conda/envs/qsiprep/lib/python3.10/site-packages/qsirecon \
/home/groups/jyeatman/software/singularity_images/qsirecon-0.23.2-unstable-25Oct24.sif \
/scratch/groups/jyeatman/howard/DBPFullwT1derivOct24 \
/scratch/groups/jyeatman/howard/DBPFullwT1derivOct24-recon/ participant --participant-label sub-$SLURM_ARRAY_TASK_ID \
-w /scratch/groups/jyeatman/howard/DBPFullwT1derivOct24-work \
--infant \
--recon-spec mrtrix_singleshell_ss3t_noACT \
--fs-license-file $GROUP_HOME/freesurferlicense.txt \
--output-resolution 2 \
-v -v
```

**Comment by tsalo** [comment]
@chiuhoward thank you for testing that. I think your error is fixed by #123.

**Comment by chiuhoward** [comment]
Thank you @tsalo I will build the latest unstable container and try again!

**Review by mattcieslak — APPROVED** [review]
This should work as a good stopgap until we have a more robust io system

**Pull request merged**